src/java.base/share/classes/jdk/internal/jimage/ImageHeader.java

```diff
@@ -180,4 +187,48 @@ int getStringsOffset() {
         return getLocationsOffset() +
                getLocationsSize();
     }
+
+    /**
+     * Tries to determine the {@link ByteOrder} of a image file present at {@code imageFile}.
+     * The implementation reads the first few relevant bytes of the image file and compares it to
+     * the expected magic bytes of a jimage. If those bytes correspond to the magic bytes, then
+     * this method returns the ByteOrder of the image file, else it returns null.
+     *
+     * @param imageFile The path to the image file
+     * @return The ByteOrder if one could be determined. Else null.
+     */
+    static ByteOrder tryDetectByteOrder(Path imageFile) {
+        byte[] bytes = new byte[Integer.SIZE]; // read only as much as the size of magic bytes
+        @SuppressWarnings("removal")
+        int numRead = AccessController.doPrivileged(
+                new PrivilegedAction<Integer>() {
+                    @Override
+                    public Integer run() {
+                        try (InputStream is = Files.newInputStream(imageFile)) {
+                            return is.read(bytes);
+                        } catch (IOException e) {
+                            // ignore
+                            return -1;
+                        }
+                    }
+                });
+        if (numRead != bytes.length) {
+            // not enough bytes to ascertain if this is a jimage
+            return null;
+        }
+        ByteBuffer bb = ByteBuffer.wrap(bytes);
+        if (bb.getInt() == MAGIC) {
+            return bb.order();
+        }
```

Comment:
I would have just added `public static final int REVERSE_MAGIC = 0xDADAFECA;` and add an else clause. to compare to REVERSE_MAGIC. Much simpler.